src/uct/ib/rc/accel/rc_mlx5_common.h

```diff
@@ -1297,4 +1301,107 @@ uct_rc_mlx5_iface_common_poll_rx(uct_rc_mlx5_iface_common_t *mlx5_common_iface,
     return count;
 }
 
+#if HAVE_IBV_EXP_DM
+/* DM memory should be written by 8 bytes (int64) to eliminate
+ * processor cache issues. To make this used uct_rc_mlx5_dm_copy_data_t
+ * datatype where first hdr_len bytes are filled by message header
+ * and tail is filled by head of message. */
+static void UCS_F_ALWAYS_INLINE
+uct_rc_mlx5_iface_common_copy_to_dm(uct_rc_mlx5_dm_copy_data_t *cache, size_t hdr_len,
+                                    const void *payload, size_t length, void *dm,
+                                    uct_ib_log_sge_t *log_sge)
+{
+    size_t head         = (cache && hdr_len) ? ucs_min(length, sizeof(*cache) - hdr_len) : 0;
+    size_t body         = ucs_align_down(length - head, sizeof(uint64_t));
+    size_t tail         = length - (head + body);
+    uint64_t *dst       = dm;
+    uint64_t padding    = 0; /* init by 0 to suppress valgrind error */
+    int i               = 0;
+
+    ucs_assert(sizeof(*cache) >= hdr_len);
+    ucs_assert(head + body + tail == length);
+    ucs_assert(tail < sizeof(uint64_t));
+
+    /* copy head of payload to tail of cache */
+    memcpy(cache->in + hdr_len, payload, head);
+
+    UCS_STATIC_ASSERT(sizeof(*cache) == sizeof(cache->out));
+    UCS_STATIC_ASSERT(sizeof(cache->in) == sizeof(cache->out));
+    UCS_STATIC_ASSERT(sizeof(log_sge->sg_list) / sizeof(log_sge->sg_list[0]) >= 2);
+
+    /* condition is static-evaluated */
+    if (cache && hdr_len) {
+        /* atomically by 8 bytes copy data to DM */
+        *(dst++) = cache->out[0];
+        *(dst++) = cache->out[1];
+        if (ucs_log_is_enabled(UCS_LOG_LEVEL_TRACE_DATA)) {
+            log_sge->sg_list[0].addr   = (uint64_t)cache;
+            log_sge->sg_list[0].length = (uint32_t)hdr_len;
+            i++;
+        }
+    }
+    if (ucs_log_is_enabled(UCS_LOG_LEVEL_TRACE_DATA)) {
+        log_sge->sg_list[i].addr   = (uint64_t)payload;
+        log_sge->sg_list[i].length = (uint32_t)length;
+        i++;
+    }
+    log_sge->num_sge = i;
+
+    /* copy payload to DM */
+    UCS_WORD_COPY(dst, payload + head, uint64_t, body);
+    if (tail) {
+        memcpy(&padding, payload + head + body, tail);
+        *(dst + (body / sizeof(uint64_t))) = padding;
+    }
```

Comment:
Do you the same doorbell interface to post it ? I guess there should memory barrier somewhere ?

Reply:
memory barrier? why? we have exclusive access to allocated buffer

Reply:
@yosefe - how it is mapped ?

Reply:
I have to take a look, but I think you are fine. I guess there is no barrier required between this write and the following doorbell record update ? 8Byte is a weird requirement.

Reply:
it's mapped by calling ibv_reg_mr with flag IBV_EXP_REG_MR_DM

Reply:
Where the destination memory resides ? on the nic ?

Reply:
yes, on card

Reply:
And it seems like you directly access the memory through the user space ? if so, it has to be mapped somehow through the kernel

Reply:
@yosefe - I'm asking since it is not clear to me how it all maps on armv8

Reply:
we are using mlx5 lib call to map HCA memory to process address space, it could be translated to MLX kernel module to attach DM.
what is the difference between ARM/PPC/Intel? I expect all these platforms works at same manner.

Reply:
Different architecture use different memory types in kernel. Is this WC memory ? I would guess so. In this context 8B write is very weird as well.

Reply:
ok, what is the better size to write?

Reply:
i think we can... it is ARM specific intrinsic?

Reply:
Actually there are x86 intrinsic for this :)

Reply:
Arm http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.den0024a/CJACGJJF.html
```They also give a hint to the memory system that caching is not useful for this data```

Reply:
x86 has 128-byte intrinsic?
is there any doc to read?

Reply:
If somebody will clearly describe what you are really trying to achieve, I will help to do it the right way :-) I also want to look at the kernel part, if it is upstream. In the past we have seen a lot if issues there as well

Reply:
I talked with @yosefe . I have a bit better understanding and @yosefe have better understanding of some of my concerns. Is this nothing especially critical, but we have to make sure that we have all the barrier for various architectures without uarch assumptions.